beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1044,33 +1044,51 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
 
     /// Returns the blobs at the given root, if any.
     ///
-    /// ## Errors
+    /// Returns `Ok(None)` if the blobs are not found. This could indicate the blob has been pruned
+    /// or that the block it is referenced by doesn't exist in our database.
     ///
-    /// May return a database error.
+    /// If we can find the corresponding block in our database, we know whether we *should* have
+    /// blobs. If we should have blobs and no blobs are found, this will error. If we shouldn't,
+    /// this will reconstruct an empty `BlobsSidecar`.
+    ///
+    /// ## Errors
+    /// - any database read errors
+    /// - block and blobs are inconsistent in the database
+    /// - this method is called with a pre-eip4844 block root
     pub fn get_blobs(
         &self,
         block_root: &Hash256,
     ) -> Result<Option<BlobsSidecar<T::EthSpec>>, Error> {
         match self.store.get_blobs(block_root)? {
             Some(blobs) => Ok(Some(blobs)),
             None => {
-                if let Ok(Some(block)) = self.get_blinded_block(block_root) {
-                    let expected_kzg_commitments = block.message().body().blob_kzg_commitments()?;
-
-                    if expected_kzg_commitments.len() > 0 {
-                        Err(Error::DBInconsistent(format!(
-                            "Expected kzg commitments but no blobs stored for block root {}",
-                            block_root
-                        )))
-                    } else {
-                        Ok(Some(BlobsSidecar::empty_from_parts(
-                            *block_root,
-                            block.slot(),
-                        )))
-                    }
-                } else {
-                    Ok(None)
-                }
+                // Check for the corresponding block to understand whether we *should* have blobs.
+                self.get_blinded_block(block_root)?
+                    .map(|block| {
+                        // If there are no KZG commitments in the block, we know the sidecar should
+                        // be empty.
+                        let expected_kzg_commitments =
+                            match block.message().body().blob_kzg_commitments() {
+                                Ok(kzg_commitments) => kzg_commitments,
+                                Err(_) => return Err(Error::BlobsUnavailable),
```

Comment:
Oh yea so I actually made this a `BlobsUnavailable` error to make it easier to handle both of those errors here: https://github.com/sigp/lighthouse/pull/3891/files#diff-707e9da29a49fe09a4bb8015297139492ad15e1663b182bbee8e0ed622d067b9R299-R315

So if we rename that error, we also need to handle that variant in the same way because in the spec if this is called with a pre-4844 block root we should respond with `ResourceUnavailable`

> Clients MUST support requesting blocks and sidecars since minimum_request_epoch, where minimum_request_epoch = max(finalized_epoch, current_epoch - MIN_EPOCHS_FOR_BLOBS_SIDECARS_REQUESTS, EIP4844_FORK_EPOCH). If any root in the request content references a block earlier than minimum_request_epoch, peers SHOULD respond with error code 3: ResourceUnavailable.

When the block doesn't have the kzg commitments field it means it's from prior to `EIP4844_FORK_EPOCH`

Reply:
Oh sorry, you already have this, I missed it. Awesome!